polygeist/tools/cgeist/Lib/Attributes.h

```diff
@@ -0,0 +1,88 @@
+//===- Attributes.h - Construct LLVMIR attributes ---------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef CGEIST_ATTRIBUTES_H
+#define CGEIST_ATTRIBUTES_H
+
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/OperationSupport.h"
+#include "llvm/IR/Attributes.h"
+
+namespace mlirclang {
+
+/// \class
+/// Facilitates the construction of LLVMIR dialect attributes for a particular
+/// argument, parameter, function, or return value.
+class AttrBuilder {
+public:
+  AttrBuilder(mlir::MLIRContext &ctx) : ctx(ctx) {}
+  AttrBuilder(const AttrBuilder &) = delete;
+  AttrBuilder(AttrBuilder &&) = default;
+
+  /// Add the LLVM attribute identified by \p kind to the builder.
+  /// Note: only unit attributes can be added using this member function.
+  AttrBuilder &addAttribute(llvm::Attribute::AttrKind kind);
+
+  /// Add the LLVM attribute identified by \p kind with a value given by \p val
+  /// to the builder.
+  AttrBuilder &addAttribute(llvm::Attribute::AttrKind kind, int64_t val);
+
+  /// Create a NamedAttribute with name \p attrName and value \p attr and add it
+  /// to the builder.
+  AttrBuilder &addAttribute(llvm::StringRef attrName, mlir::Attribute attr);
+
+  /// Add the given named attribute \p attr to the builder.
+  AttrBuilder &addAttribute(mlir::NamedAttribute attr) {
+    attrs.set(attr.getName(), attr.getValue());
+    return *this;
+  }
+
+  /// Add the LLVM attribute identified by \p kind to the builder "passthrough"
+  /// named attribute.
+  AttrBuilder &addPassThroughAttribute(llvm::Attribute::AttrKind kind);
+
+  /// Add the given attribute \p attr to the builder "passthrough" named
+  /// attribute.
+  AttrBuilder &addPassThroughAttribute(mlir::Attribute attr);
+
+  /// Return true if the builder has the specified attribute.
+  bool contains(llvm::StringRef attrName) const;
```

Comment:
Would it make sense to have a second overload of this function with parameter type `llvm::Attribute::AttrKind`, so users of this API don't need to retrieve the name first? 

Same goes for `getAttr`.

Reply:
good idea, done.